src/core/search/ast_expr.h

```diff
@@ -4,21 +4,86 @@
 
 #pragma once
 
+#include <algorithm>
+#include <iostream>
+#include <memory>
 #include <ostream>
+#include <vector>
 
 namespace dfly {
 
 namespace search {
 
-class AstExpr {};
+// Describes a single node of the filter AST tree.
+class AstNode {
+ public:
+  virtual ~AstNode() = default;
+
+  // Check if this input is matched by the node.
+  virtual bool Check(std::string_view input) const = 0;
+
+  // Debug print node.
+  virtual std::string Debug() const = 0;
+};
+
+using NodePtr = std::shared_ptr<AstNode>;
+using AstExpr = NodePtr;
+
+template <typename T, typename... Ts> AstExpr MakeExpr(Ts... ts) {
+  auto ptr = std::make_shared<T>(std::forward<Ts>(ts)...);
+  return std::dynamic_pointer_cast<AstNode>(move(ptr));
```

Comment:
`shared_ptr`s allow up-cast, so as long as `T` inherits from `AstNode`, you don't need this `dynamic_pointer_cast`. Am I missing something?

Example:

```c++
struct Parent { virtual ~Parent() = default; };
struct Derived : Parent {};

int main() {
  std::shared_ptr<Parent> p = std::make_shared<Derived>();
}
```

Reply:
Yes, my bad